src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/Orleans3CompatibleStringKeyHasher.cs

```diff
@@ -37,17 +38,63 @@ public int Hash(byte[] data)
             // PickHasher parameters are the same for both calls so we need to analyze data content to distinguish these cases.
             // It doesn't word if string key is equal to grain type name, but we consider this edge case to be negligibly rare.
 
-            // reducing allocations if data is not a grain type
-            if (data.Length >= _grainType.Length && Encoding.UTF8.GetByteCount(_grainType) == data.Length)
+            if (IsGrainTypeName(data))
+                return _innerHasher.Hash(data);
+
+            var extendedLength = data.Length + 8;
+            if (extendedLength <= 256)
+            {
+                Span<byte> extended = stackalloc byte[extendedLength];
```

Comment:
Stack-allocated should be to a constant of power of 2, then sliced. It's cheaper.
The extra space (extended) should also be cleared, just to be safe (see SkipLocalsInit).